### ISSUE TITLE
why new weights promoted speed up

### ISSUE THREAD

**Issue by l1t1** [gcp/leela-zero#1978]
recent four weights winrates are larger, is it related with new version client?

**Comment by wonderingabout** [comment]
i dont think so
just a guess, but it may be because our leela-zero is finally overcoming the elf fuseki (4-4 quadruple approach), in favour of 3-3 early

**Comment by Borthralla** [comment]
I'm trying it out on CGOS and it seems to be doing pretty well! :)
http://www.yss-aya.com/cgos/19x19/cross/LZ-185-1080ti.html

**Comment by Ishinoshita** [comment]
Under match condition, new version should make both black and white stronger, so no difference.

New version 'just' speed up game generation rate (great achievement in fact!), but on training side, scale is still the same, number of self-play game (training window size, learning steps, Self-play Elo gragh). If I got it correct, new version is just shrinking time scale.

So I wouldn't expect LZ to speed up just under the effect of new version (on a self-play games scale). More because she has moved away faster from some local minima, just by herself or due to some tweak in the self-play parameters.

The remarkable part is that, IIRC, the training window has been reset with LZ185 promotion. That mean LZ should now trains on its own recent games, and is still moving fast. So those new games maybe of better quality. Some also noticed that she has changed her opening, rediscovering the virtue of early 3-3 invasion. Let's hope she is not just moving in circles. Would be interedting, after a while, to benchmark this supposed recent strength increase by matching latest net with older net version.

**Comment by gcp** [comment]
>The remarkable part is that, IIRC, the training window has been reset with LZ185 promotion

It resets after finishing the 256k steps net, not after a promotion. The training machine doesn't know a net promotes (that happens quite a while after it submits it), it just notices at the next cycle that when it downloads best-network that it's something else.

> @gcp said he set things to randomly sample 25% of Elf1 games, IIRC meaning that randomly sampled Elf1 games are added to the training window up to amount to 25%.

I tried this again, this time randomly sampling 50% (i.e. cutting out half of the) ELFv1 games, and replacing them by regular games. It's probably that which is now working, and causing the style evolution.

It looks like we will be able to drop ELF games entirely soon. At least I will try this soon.

**Comment by Friday9i** [comment]
At fixed 1600 visits, used for selfplay, LZ186 is significantly stronger than ELFv1 : Elfv1 needs about 2200 visits against LZ181 with 1600 visits to get 50% winrate, (source: #1914) so, against LZ186, Elfv1 probably needs 2500 or 3000 visits (I guess it's around 150 Elo difference).
That demonstrates selfplay games from latest LZ are significantly better than from Elfv1: hence, we should probably remove all Elfv1 games from training sets, asap 😉

**Comment by roy7** [comment]
The problem is if the training targets radically change too much from the current network, it will be hard to get a new network that can exceed the 55% win rate cutoff. Slowly easing off of the ELF might work better. Each new network that promotes with less ELF games will be that much reliant on ELF data.

If we can get another promotion or two then maybe we east back on ELF another portion? And do that a few times until it's at 0%. Vs just yanking it all out at once and having a network with lots of ELF knowledge trying to learn from zero ELF games.

I updated RoyalZeroSlow on OGS to the latest 40b network and it's 3-0 playing against my RoyalLZ-ELF bot. Been fun games. Now RZS is giving LZ-ELF an extra stone handicap to see what happens. ;) Game not over yet, but RZS has crawled from 13.9% at game start to 36.1% now in mid game. https://online-go.com/game/15112419

**Comment by Friday9i** [comment]
I think we should try to partially reduce random play, eg -m 30 but also randomvisits 50 or 100 for a % of the games
However, it's a pity that some people do it by manipulating selfplay games parameters... Didn't know it represents as much as 40% of the games, that's insane!

**Comment by wonderingabout** [comment]
somehow my comment disappeared but 
@Ishinoshita leela zero is an ai, i prefer to call it "it"

**Comment by alreadydone** [comment]
@Friday9i People experimented with --softmax-temp, --randomtemp, and introduced blunder threshold in https://github.com/sai-dev/sai/issues/8. I think -m 150~200 for LZ 19x19 should be fine (and enough to save time in no-resign games), but since the majority is still on -m 999 I'd rather go back to AGZ's -m 30.
@wonderingabout Leela is a girl though.

**Comment by Ishinoshita** [comment]
On the same line, and just as a thought, as I came across this report from @sethtroisi about an experiment on minigo:
>  [Experiment - Sensitivity to mislabeled result data](https://github.com/tensorflow/minigo/issues/483)
*Experiment
Load 10 "golden chunks" (v9 chunk 250 to 259), this is 20 million examples taken from ~600,000 selfplay games from strong v9 models.When training flip the result (from white to black win and vice versa) in some percent {1, 2, 5, 10} of examples.Train really fast on a bunch of TPUs at a couple of different network sizes.
*Results
TL;DR Flipping 1% and 2% of results doesn't have much impact, 5% and 10% have a much big impact on confidence and value accuracy. 

I initially made a link with the question of what fraction of fake data/games could derail LZ training. But it's in fact quite different as, here above, policy target is unaffected, only value suffers from noise injection. Nonetheless, I'm wondering about a possible connection with the use of randomess in LZ.

* With -m 30 (AGZ), game opening is randomized,; it may happen that a complete blunder seals the game early, but then, both side will play their best moves and the outcome will be a 'good' evaluation for all positions after 30 move count, under the current policy. Most positions will receive the 'right' label. Exploration is limited though, to moves that Dirichlet noise pushes high enough and that are more or less agreed/dismissed by the search. Seldom, a noise-promoted moves will prevail and bend the game trajectory. If I got it correct, minigo use an equvallent temperature_cutoff of 30 moves.

* With -m 999, either side may play many suboptimal moves or even a total blunder, all along the game, that may reverse the game outcome, possibly several times in the game (that happens, from what I have observed). So final outcome is a much more noisy training target. It can be a seen as a 'flipped' target for a fraction of the positions in the game, amounting to minigo's expriment. On the other hand, exploration is boosted, as T=1 will randomize the game trajectory all along, maximizing state space exploration.

There must be some trade-off here. I know that AZ paper reports -m999 but we badly lack some details on AZ overall. I'm eager to see some experimentation in this respect in the future. Once we get rid if elf mixing and get a stable baseline. And that things can be tried 'all things being equal' ;-)

**Comment by l1t1** [comment]
it seems 40b weights are on their way now

**Comment by gcp** [comment]
>What makes Leela Zero learn better from its own games as opposed to ELF V1?

The ELF games are fixed, i.e. they do not evolve at all. The games from LZ's networks allow it to see the mistakes from its current network.

**Comment by Friday9i** [comment]
@gcp When do you think you will entirely remove Elf games from training? Would you mind to post something (on sjeng.org or here?) to inform us of that change? 
(and once again, thanks for this wonderful project!)

**Comment by jillybob** [comment]
LZ networks seem stalled now, can we remove more ELF games?

**Comment by l1t1** [comment]
elf is still stronger than lz 188 in middle and endgame.

**Comment by wonderingabout** [comment]
> 
> 
> elf is still stronger than lz 188 in middle and endgame.

yes, facebook's work on the 20 block training is impressive

i cant even begin to imagine how scary alphago's refined 40 block is

**Comment by gcp** [comment]
FWIW ELF games were completely removed a few days ago.
